### DISCUSSION TITLE
Use a current "set" of units for UI uniformity

### DISCUSSION THREAD

**Discussion by wiwalsh** [angularsen/UnitsNet#869] [General]
When interacting with a user interface it would be nice to have a way to get the numerical value of a generic unit by passing a unit setting container.  An example use would be this below.  Here in this generic, the SelectedUnits would have settings for each specific unit type.  This would enable this list to be modified by the user and have all locations where that unit is use to be updated and have the values updated at the same time throughout the program (or on a specific screen or table if desired).  The class SelectedUnits would be just a list of enums that each specific unit class would pick out the appropriate enum and return a double.

Does something like this already exist or is it something that can be added?

```c#
 public class Constraint<T>
  {
      public T Max { get; private set; }
      public T Min { get; private set; }
      public string Name { get; private set; }

      public double getMax(SelectedUnits units)
      {        
          return Max(units);
      }
      public Constraint(string name, T min, T max)
      {
          this.Name = name;
          this.Min = min;
          this.Max = max;
      }
    }
```

**Comment by angularsen** [comment]
Hi, so if I understand correctly you would have something like this:

```c#
var selectedUnits = new Dictionary<string, Enum> 
{
    { "Length", LengthUnit.Centimeter },
    { "Mass", MassUnit.Kilogram },
    // etc.
};

var maxLength = Length.MaxValue.As((LengthUnit)selectedUnits["Length"]); // Max length in centimeters.
```

There is no such concept in the library today, but there is work in progress to redesign our `UnitSystem` to support something similar to what you describe:
https://github.com/angularsen/UnitsNet/pull/709

This is a breaking change and would have to be part of #563 .

**Reply by wiwalsh** [reply]
I think so is the answer.  I've created a somewhat kludgy workaround using extension methods.  I realize this is messy, but it works.  How can I help make this a reality in UnitsNet?

implementation in a class
 public class Constraint<T> where T : UnitsNet.IQuantity
    {
      public T Max { get; private set; }
      public T Min { get; private set; }
      public string Name { get; private set; }

      public double getMax(SelectedUnits units)
      {
          return Max.ToValue(units);
      }

      public Constraint(string name, T min, T max)
      {
          this.Name = name;
          this.Min = min;
          this.Max = max;
      }
    }

container class for selected units (could be a dictionary too)
    public class SelectedUnits
    {
        public AngleUnit Angle;
        public LengthUnit Length;

        public SelectedUnits()
        {
            Angle = AngleUnit.Degree;
            Length = LengthUnit.Inch;
        }
    }

extension methods that make it work

 public static double ToValue(this IQuantity q, SelectedUnits u) 
        {
            
            if (q.GetType() == typeof(Angle))
            {
                return ToValue((Angle) q, u);
            }
            else if (q.GetType() == typeof(Length))
            {
                return ToValue((Length)q, u);
            }
            else
            {
                return double.MinValue;
            }
            
        }
        public static double ToValue(this Angle a, SelectedUnits u)
        {
            return a.ToUnit(u.Angle).Value;
        }
        public static double ToValue(this Length a, SelectedUnits u)
        {
            return a.ToUnit(u.Length).Value;
        }

**Reply by angularsen** [reply]
I think for now, you are best off implementing this in your own application.
If you want to help move this idea into the library, I suggest you contribute to #709 .
